Lib/tarfile.py

```diff
@@ -349,7 +349,7 @@ def __init__(self, name, mode, comptype, fileobj, bufsize):
             fileobj = _StreamProxy(fileobj)
             comptype = fileobj.getcomptype()
 
-        self.name     = name or ""
+        self.name     = os.path.abspath(name) if name else ""
```

Comment:
`os.fspath` looks like a good solution to me. Please add a test in `Lib/test/test_tarfile.py` that opens `gzipname` with mode 'w|gz' flag to make sure there is no `AttributeError`

Reply:
@tirkarthi after opening `gzipname` with mode `w|gz` all remaining tests, that need to extract data from `gzipname` will not pass, because opening rewrites file content.

I added test to `StreamWriteTest` class, where I open for write `tmpname`, because it's designed for writing and insensitive to content change. Also, since this test inside of `StreamWriteTest`, opening by path-like filename will be performed for each mode (gz, lz, bz).